server/chat.ts

```diff
@@ -233,142 +239,114 @@ Chat.namefilter = function (name, user) {
-Chat.translations = new Map();
+// language id -> language name
+const languages: Map<string, string> = new Map();
+// language id -> (english string -> translated string)
+const translations: Map<string, Map<string, [string, string[], string[]]>> = new Map();
 
+// tslint:disable-next-line: no-floating-promises
 FS(TRANSLATION_DIRECTORY).readdir().then(files => {
-	for (let fname of files) {
+	for (const fname of files) {
 		if (!fname.endsWith('.json')) continue;
 
-		/**
-		 * @type {{name: string?, strings: {[k: string]: string}?}} content
-		 */
-		const content = require(`../${TRANSLATION_DIRECTORY}${fname}`);
+		interface TRStrings {
+			[k: string]: string;
+		}
+		const content: {name: string, strings: TRStrings} = require(`../${TRANSLATION_DIRECTORY}${fname}`);
 		const id = fname.slice(0, -5);
 
-		Chat.languages.set(id, content.name || "Unknown Language");
-		Chat.translations.set(id, new Map());
+		languages.set(id, content.name || "Unknown Language");
+		translations.set(id, new Map());
 
 		if (content.strings) {
 			for (const key in content.strings) {
-				/** @type {string[]} */
-				const keyLabels = [];
-				/** @type {string[]} */
-				const valLabels = [];
-				const newKey = key.replace(/\${.+?}/g, string => {
-					keyLabels.push(string);
+				const keyLabels: string[] = [];
+				const valLabels: string[] = [];
+				const newKey = key.replace(/\${.+?}/g, str => {
+					keyLabels.push(str);
 					return '${}';
 				}).replace(/\[TN: ?.+?\]/g, '');
-				const val = content.strings[key].replace(/\${.+?}/g, string => {
-					valLabels.push(string);
+				const val = content.strings[key].replace(/\${.+?}/g, (str: string) => {
+					valLabels.push(str);
 					return '${}';
 				}).replace(/\[TN: ?.+?\]/g, '');
-				Chat.translations.get(id).set(newKey, [val, keyLabels, valLabels]);
+				translations.get(id)!.set(newKey, [val, keyLabels, valLabels]);
 			}
 		}
 	}
 });
 
-/**
- * @param {string?} language
- * @param {TemplateStringsArray | string} strings
- * @param {any[]} keys
- */
-Chat.tr = function (language, strings, ...keys) {
+function tr(language: string | null, strings: TemplateStringsArray | string, ...keys: any[]) {
 	if (!language) language = 'english';
 	language = toID(language);
-	if (!Chat.translations.has(language)) throw new Error(`Trying to translate to a nonexistent language: ${language}`);
+	if (!translations.has(language)) throw new Error(`Trying to translate to a nonexistent language: ${language}`);
 	if (!strings.length) {
-		// @ts-ignore no this isn't any type
-		return (function (strings, ...keys) {
-			Chat.tr(language, strings, ...keys);
+		return ((fStrings: TemplateStringsArray | string, ...fKeys: any) => {
+			tr(language, fStrings, ...fKeys);
 		});
 	}
 
 	// If strings is an array (normally the case), combine before translating.
-	let string = Array.isArray(strings) ? strings.join('${}') : strings;
+	const trString = Array.isArray(strings) ? strings.join('${}') : strings as string;
```

Comment:
Does this actually need to be cast? I thought TS supported doing it automatically.

Reply:
according to me deleting the cast and checking, yes